File: 4-sumcheck/python/verifier.py
from random import randrange
from typing import Callable
from utils import deg_j


class Verifier:
    def __init__(self, g, g_arity, H) -> None:
        """Initialize verifier with the claimed witness H"""
        self.g = g
        self.g_arity = g_arity
        self.H = H
        self.random_challenges = []
        self.round = 1
        self.polynomials = []

    def receive_polynomial(self, polynomial: Callable):
        """receive the latest polynomial from the Prover"""
        self.polynomials.append(polynomial)

    def check_latest_polynomial(self):
        """Verify that self.latest_polynomial is a univariate polynomial of at most deg_j(g) and 
        that g_{j-1}(r_{j-i}) = g_j(0)+g_j(1)"""
        latest_poly = self.polynomials[-1]

        # print("debug %s" % len(self.polynomials))
        # print("round", self.round, "poly:", latest_poly(0), latest_poly(1))
        deg_latest = deg_j(latest_poly, 0)
        deg_max = deg_j(self.g, self.round-1)
        if deg_latest > deg_max:
            raise ValueError("Prover sent polynomial of degree {} greater than expected: {}".format(
                deg_latest, deg_max))

        new_sum = latest_poly(0)+latest_poly(1)
        if self.round == 1:
            check = self.H
        else:
            check = self.polynomials[-2](self.random_challenges[-1])
            # print("check", check)
        if check != new_sum:
            raise ValueError(
                "Prover sent incorrect polynomial: {},expected: {}".format(new_sum, check))

    def get_new_random_value_and_send(self, p):
        """Get a new random value, append it to random challenges, send it to the prover, and update the round"""
        self.random_challenges.append(randrange(2))
        p.receive_challenge(self.random_challenges[-1])
        self.round += 1

    def evaluate_and_check_g_v(self):
        assert len(self.random_challenges) == self.g_arity-1
        self.random_challenges.append(randrange(2))
        g_final = self.g(*self.random_challenges)
        check = self.polynomials[-1](self.random_challenges[-1])

        if g_final != check:
            raise ValueError(
                "Prover sent incorrect final polynomial: {},expected: {}".format(g_final, check))
        else:
            print("VERIFIER ACCEPTS")
            return True

File: 4-sumcheck/src/verifier.rs
use crate::{prover::Prover, utils::FArity};

pub(crate) struct Verifier {
    g: FArity,
    h_claim: usize,
    random_challenges: Vec<usize>,
    round: usize,
    second_last_poly: Option<FArity>,
    last_polynomial: Option<FArity>,
}
impl Verifier {
    pub(crate) fn new(g: FArity, h_claim: usize) -> Self {
        Self {
            g,
            h_claim,
            random_challenges: vec![],
            round: 1,
            second_last_poly: None,
            last_polynomial: None,
        }
    }

    pub(crate) fn receive_polynomial(&mut self, polynomial: FArity) {
        self.second_last_poly = self.last_polynomial.take();
        self.last_polynomial = Some(polynomial);
    }

    pub(crate) fn check_latest_polynomial(&self) {
        let last_polynomial = self.last_polynomial.as_ref().unwrap();
        let deg_latest = last_polynomial.deg_j(0);
        let deg_max = self.g.deg_j(self.round - 1);

        if deg_latest > deg_max {
            panic!(
                "Prover sent polynomial of degree {} greater than expected {}",
                deg_latest, deg_max
            );
        }
        let a = last_polynomial.call_f(vec![0]);
        let b = last_polynomial.call_f(vec![1]);
        dbg!(a,b);

        let new_sum = last_polynomial.call_f(vec![0]) + last_polynomial.call_f(vec![1]);
        let check = match self.round {
            1 => self.h_claim,
            _ => self
                .second_last_poly
                .as_ref()
                .unwrap()
                .call_f(vec![*self.random_challenges.last().unwrap()]),
        };
        if check != new_sum {
            panic!(
                "Prover sent incorrect polynomial value: {}, expected: {}",
                new_sum, check
            );
        }
    }

    pub(crate) fn get_new_random_value_and_send(&mut self, p: &mut Prover) {
        let new_challenge: usize = rand::random::<bool>().into();
        self.random_challenges.push(new_challenge);
        p.receive_challenge(new_challenge);
        self.round += 1;
        println!("VERIFIER: advancing to round {}", self.round);
    }

    pub(crate) fn evaluate_and_check_g_v(&mut self) {
        println!("VERIFIER: final round");
        let new_challenge: usize = rand::random::<bool>().into();
        self.random_challenges.push(new_challenge);
        let g_final = self.g.call_f(self.random_challenges.clone());
        let check = self
            .last_polynomial
            .take()
            .unwrap()
            .call_f(vec![*self.random_challenges.last().unwrap()]);
        if g_final != check {
            panic!(
                "Prover sent incorrect final polynomial value: {}, expected: {}",
                g_final, check
            );
        } else {
            println!("VERIFIER ACCEPTS")
        }
    }
}
